components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/OAuthAppTenantResolverValve.java

```diff
@@ -0,0 +1,142 @@
+ * qualified urls are disabled, client id is unique across the server.
+ */
+public class OAuthAppTenantResolverValve extends ValveBase {
+
+    private static final Log log = LogFactory.getLog(OAuthAppTenantResolverValve.class);
+
+    @Override
+    public void invoke(Request request, Response response) throws IOException, ServletException {
+
+        try {
+            if (!IdentityTenantUtil.isTenantQualifiedUrlsEnabled() && isOAuthRequest(request)) {
+                String appTenant = "";
+                String clientId = request.getParameter(CLIENT_ID);
+
+                if (StringUtils.isEmpty(clientId) && isOAuth10ARequest(request)) {
+                    clientId = request.getParameter(OAUTH_CONSUMER_KEY);
+                }
+
+                // If empty, try to get the client id from the authorization header.
+                if (StringUtils.isEmpty(clientId)) {
+                    try {
+                        String[] credentials = OAuth2Util.extractCredentialsFromAuthzHeader(request);
+                        if (credentials.length == 2 && StringUtils.isNotEmpty(credentials[0])) {
+                            clientId = credentials[0];
+                        }
+                    } catch (OAuthClientAuthnException e) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Error while extracting credentials from authorization header.", e);
+                        }
+                    }
+                }
+
+                if (StringUtils.isNotEmpty(clientId)) {
+                    try {
+                        OAuthAppDO oAuthAppDO = OAuth2Util.getAppInformationByClientIdOnly(clientId);
+                        appTenant = OAuth2Util.getTenantDomainOfOauthApp(oAuthAppDO);
+                    } catch (IdentityOAuth2Exception e) {
+                        log.error("Error while getting oauth app for client Id: " + clientId, e);
+                    } catch (InvalidOAuthClientException e) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Error while getting oauth app for client Id: " + clientId, e);
+                        }
+                    }
+                }
+
+                // Set the tenant name to thread local properties.
+                if (StringUtils.isNotEmpty(appTenant)) {
+                    IdentityUtil.threadLocalProperties.get().put(TENANT_NAME_FROM_CONTEXT, appTenant);
+                }
+            }
+            getNext().invoke(request, response);
+        } finally {
+            // Clear thread local tenant name.
+            unsetThreadLocalContextTenantName(request);
+        }
+    }
+
+    /**
+     * Check whether the request is an OAuth request.
+     *
+     * @param request Http servlet request.
+     * @return True if the request is an OAuth request.
+     */
+    private boolean isOAuthRequest(Request request) {
```

Comment:
- When hosting IS in a proxy context (Ex: example.com/idp/carbon) starts with check may not work.
- Retrieve identity server url using util method and build the absolute url using url builder (check for a util method)
- Check request uri starts with absolute url

Reply:
TODO::